Correctly clear PriorityValue local value

grokys opened the pull request:
## What does the pull request do?

Fix for #8372: `_localValue` was not being cleared in `PriorityValue`.

However this exposed another problem with reentrancy in `PropertySetterInstance.Dispose`, which needed fixing too. See [the commit](https://github.com/AvaloniaUI/Avalonia/pull/8373/commits/857bfb5bd2b863825c09fbfb780dc379fba4d345) for the slightly confusing explanation.

## Checklist

- [x] Added unit tests (if possible)?

## Fixed issues

Fixes #8372

grokys committed:
Add failing test for #8372.

grokys committed:
Correctly clear local value in PriorityValue.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021296-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

grokys commented:
Marking this as a draft because fixing this has exposed another problem.

grokys committed:
Added additional failing test.

Exposed by the previous fix for #8372: re-entrancy in `PropertySetterInstance.Dispose()` is causing detaching a style to call `ClearValue` on the property. Previously this wasn't a problem as `ClearValue` didn't work, but now it is.

(Also added one passing test which tests the same scenario in `PropertySetterBindingInstance` for future coverage)

grokys committed:
Prevent re-entrancy in PropertySetterInstance.Dispose.

The call to `_subscription.Dispose()` causes `BindingEntry.Dispose()` to call `_subscription.Dispose()`, but in this case the `BindingEntry._subscription` instance is the `PropertySetterInstance`!

Except now `PropertySetterInstance._subscription` is null, and so `PropertySetterInstance.Dispose` called `ClearValue`, which is obviously wrong.

grokys commented:
Additional issue now fixed.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0021307-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]
